src/fits.jl

```diff
@@ -0,0 +1,153 @@
+# helper functions
+
+# Gets the data from hdu
+function getdata(hdu::ImageHDU)
+    data = read(hdu)
+    d = ndims(data)
+    return permutedims(data, d:-1:1)
+end
+
+#------------------------------------------------------------------
+
+# FITS interface for basic reduction methods
+
+# code generated with codegen
+for func in (:flat_correct, :subtract_bias)
+    @eval $func(frame::ImageHDU, correction::AbstractArray; kwargs...) = $func(getdata(frame), correction; kwargs...)
+    @eval $func(frame::AbstractArray, correction::ImageHDU; kwargs...) = $func(frame, getdata(correction); kwargs...)
+    @eval $func(frame::ImageHDU, correction::ImageHDU; kwargs...) = $func(getdata(frame), correction; kwargs...)
+    @eval $func(frame::String, correction; hdu = 1, kwargs...) = $func(FITS(frame)[hdu], correction; kwargs...)
+    @eval $func(frame, correction::String; hdu = 1, kwargs...) = $func(frame, FITS(correction)[hdu]; kwargs...)
+    @eval begin
+        function $func(frame::String, correction::String; hdu = (1, 1), kwargs...)
+            hdus = hdu isa Integer ? (hdu, hdu) : hdu
+            return $func(FITS(frame)[hdus[1]], FITS(correction)[hdus[2]]; kwargs...)
+        end
+    end
+end
+
+for func in (:flat_correct!, :subtract_bias!)
+    @eval $func(frame::AbstractArray, correction::ImageHDU; kwargs...) = $func(frame, getdata(correction); kwargs...)
+    @eval $func(frame::AbstractArray, correction::String; hdu = 1, kwargs...) = $func(frame, FITS(correction)[hdu]; kwargs...)
+end
+
+for func in (:crop, :trim, :subtract_overscan)
+    @eval $func(frame::ImageHDU, args...; kwargs...) = $func(getdata(frame), args...; kwargs...)
+    @eval $func(filename::String, args...; hdu = 1, kwargs...) = $func(FITS(filename)[hdu], args...; kwargs...)
+end
+
+trim(frame::ImageHDU, idxs::Symbol) = trim(frame, read_header(frame)[string(idxs)])
+subtract_overscan(frame::ImageHDU, key::Symbol; kwargs...) = subtract_overscan(frame, read_header(frame)[string(key)]; kwargs...)
+
+
+# Documentation for code generated with codegen
+"""
+    crop(::FITSIO.ImageHDU, shape; force_equal = true)
+    crop(filename, shape; hdu=1, force_equal = true)
+
+Load a FITS file or HDU before cropping.
+"""
+function crop end
+
+
+"""
+    trim(::FITSIO.ImageHDU, idxs)
+    trim(filename, idxs; hdu=1)
+
+Load a FITS file or HDU before trimming. If `idxs` is a symbol it will be read from the FITS header with that key (case sensitive).
+"""
+function trim end
+
+
+"""
+    flat_correct(frame, flat_frame; [hdu = 1], norm_value = mean(flat_frame))
```

Comment:
Here and also in other places below: I don't think you should use `[...]` around the keyword arguments in the docstrings.  It usually indicates an optional argument, but keyword arguments are always optional.  This is also used in an inconsistent way, only around `hdu`

Reply:
Here `hdu` is an optional keyword argument, as in at times it is possible that the constructor `flat_frame` will not have `hdu` as a keyword argument. `hdu` is only present in the constructor when `frame`, `flat_frame` or both are passed as strings, otherwise it is not present as a keyword argument. So, I think it is correct to keep `hdu` as `[hdu = 1]`, what's your take on this?

Reply:
Ok, I see what you mean, then it's probably fine

Reply:
what's awkward here, is that this docstring doesn't match other, similar docstrings. There already exists a docstring for `flat_correct` that defines most of the function. See below how `subtract_overscan` uses a different docstring?
```julia
    subtract_overscan(::FITSIO.ImageHDU, idxs; [dims])
    subtract_overscan(filename, idxs; hdu=1, [dims])
```
why not switch to that for consistency?

Reply:
If we write out all variable types it takes in, it becomes quite clumsy. You suggested to switch for `subtract_bias` as (https://github.com/JuliaAstro/CCDReduction.jl/pull/18#discussion_r436244444) so I switched here as well. Thoughts?

Reply:
Should we switch to those verbose docstrings with all types explicitly mentioned?

Reply:
I think for now it's nice to show the different docstrings for ImageHDU and strings because they add functionality of choosing the hdu to load and sometimes using  header keys directly with arguments. I think this would work out easier in a separate docstring for now. In the future, we could try to find a way to succinctly incorporate all the documentation in a unified way, but I think this is easier for us and provides the right amount of granularity for users.